text/3537-msrv-resolver.md

```diff
@@ -0,0 +1,879 @@
+
+When `Cargo.lock` is resolved for latest dependencies, independent of MSRV,
+verifying the MSRV becomes difficult as they  must either juggle two lockfiles, keeping them in sync, or use the unstable `-Zminimal-versions`.
+The two lockfile approach also has all of the problems shown earlier in writing the lockfile.
+
+When only keeping MSRV-incompatible `dev-dependencies`,
+one lockfile can be used but it can be difficult to edit the `Cargo.lock` to ensure you get new `dev-dependencies` without infecting other dependency types.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## The `rust-version` field
+
+*(update to [manifest documentation](https://doc.rust-lang.org/cargo/reference/manifest.html#the-rust-version-field))*
+
+The `rust-version` field is an optional key that tells cargo what version of the
+Rust language and compiler your package can be compiled with. If the currently
+selected version of the Rust compiler is older than the stated version, cargo
+will exit with an error, telling the user what version is required.
+To support this, Cargo will prefer dependencies that are compatible with your `rust-version`.
+
+The first version of Cargo that supports this field was released with Rust 1.56.0.
+In older releases, the field will be ignored, and Cargo will display a warning.
+
+```toml
+[package]
+# ...
+rust-version = "1.56"
+```
+
+The Rust version must be a bare version number with two or three components; it
+cannot include semver operators or pre-release identifiers. Compiler pre-release
+identifiers such as -nightly will be ignored while checking the Rust version.
+The `rust-version` must be equal to or newer than the version that first
+introduced the configured `edition`.
+
+The `rust-version` may be ignored using the `--ignore-rust-version` option.
+
+Setting the `rust-version` key in `[package]` will affect all targets/crates in
+the package, including test suites, benchmarks, binaries, examples, etc.
+
+## Rust Version
+
+*(update to [Dependency Resolution's Other Constraints documentation](https://doc.rust-lang.org/cargo/reference/resolver.html))*
+
+When multiple versions of a dependency satisfy all version requirements,
+cargo will prefer those with a compatible `package.rust-version` over those that
+aren't compatible.
+Some details may change over time though `cargo check && rustup update && cargo check` should not cause `Cargo.lock` to change.
+
+#### `resolver.precedence`
+
+*(update to [Configuration](https://doc.rust-lang.org/cargo/reference/config.html))*
+
+* Type: string
+* Default: "rust-version"
+* Environment: `CARGO_RESOLVER_PRECEDENCE`
+
+Controls how `Cargo.lock` gets updated on changes to `Cargo.toml` and with `cargo update`.  This does not affect `cargo install`.
+
+* `maximum`: Prefer the highest compatible versions of dependencies
+* `rust-version`: Prefer dependencies where their `rust-version` is compatible with `package.rust-version`
```

Comment:
> > rust-version: Prefer dependencies where their rust-version is compatible with package.rust-version
> > rust-version: what is defined in the package (default for v3 resolver)
> 
> This description in the text is not clear to me. What does it mean? Perhaps some basic examples of what a `cargo update` would do with a tiny dependency change depending on `clap_derive.rust-version`, current `package.rust-version`, and `rustc --version`, and the different resolver strategies would be beneficial.
> 
> For instance, if I have:
> 
>     * `rustc --version` == `1.71`
> 
>     * `package.rust-version` == `1.60`
> 
> 
> What does `cargo update` do? Does it try to update my dependencies to those compatible with 1.71? Or only to those that list `1.60` as their _minimum_ version? If the former, would that mean that while when I published I'm promising it works with 1.60, I by default won't be testing things out with newer versions?
> 
> Again, whatever the answer is here, I wish it were in the RFC text.

Reply:
I've updated that line to be
> * `rust-version`: prefer dependencies where their `package.rust-version` is greater than or equal to your `package.rust-version`


Note that this is in the Guide section and is written specifically to go into cargo's documentation.  We go into more specifics in the [reference section](https://github.com/epage/rfcs/blob/msrv/text/3537-msrv-resolver.md#cargo-resolver) which I hope will make this more clear.

Reply:
I still don't understand what it actually does. Would a table of examples help generally?